docs/extended_api/asynchronous_operations_library/aligned_size_t.md

```diff
@@ -0,0 +1,61 @@
+---
+grand_parent: Extended API
+parent: Asynchronous operations library
+---
+
+# cuda::**aligned_size_t**
+
+Defined in header [`<cuda/barrier>`](../headers/barrier.md)
+
+Defined in header [`<cuda/pipeline>`](../headers/pipeline.md)
+
+```c++
+template<size_t Alignment>
+struct aligned_size_t;
+```
+
+The class template `cuda::aligned_size_t` is a _shape_ representing an extent of bytes with a statically defined (address and size) alignment.
+
+## Template parameters
+
+| Alignment | the address & size alignement of the byte extent |
+
+## Data members
+
+| [align](./aligned_size_t/align.md) | the alignment of the byte extent |
+| [value](./aligned_size_t/value.md) | the size of the byte extent      |
+
+## Member functions
+
+| [(constructor)](./aligned_size_t/constructor.md) | constructs an _aligned size_                                                      |
+| (destructor) [implicitly declared]               | trivial implicit destructor                                                       |
+| operator= [implicitly declared]                  | trivial implicit copy/move assignment                                             |
+| operator size_t                                  | implicit conversion to [`size_t`](https://en.cppreference.com/w/cpp/types/size_t) |
+
+## Notes
+
+If `value` is not a multiple of `align` the behavior is undefined.
+
+If `Alignment` is not a [valid alignment](https://en.cppreference.com/w/c/language/object#Alignment) the behavior is undefined.
+

```

Comment:
How does this interact with the "validity" requirement of `memcpy_async` ?

For example, here:

```c++
T a, b;
cuda::memcpy_async(g, &b, &a, cuda::aligned_size_t<alignof(T) * 2>(0), b);
assert(&a %  (alignof(T) * 2) == 0);
```

the application is stating that the alignment of the `&a` and `&b` pointers is `2 * alignof(T)`, which might or might not be the case depending on what addresses these variables have at run-time (if any). Is this stating that such an alignment requirement is required for those addresses to be valid? 

That is, would it be a sound program transformation to remove the `assert` that (hopefully) is checking that this is indeed the case since we are stating that it is?

Reply:
Correct. If the user uses aligned_size_t, they are providing an assumption that the pointers (and the size) are aligned at least to that alignment, and if that's not the case, it's on the user (and UB).

Reply:
Thanks. Should we call this out a bit more explicitly somewhere ?

Reply:
Yeah, we should. Good catch. I thought we did.

cc @c0riolis

Reply:
Done, thanks.